dev/src/bulk-writer.ts

```diff
@@ -714,27 +748,34 @@ export class BulkWriter {
     const tag = requestTag();
     const pendingBatch = this._bulkCommitBatch;
 
-    this._bulkCommitBatch = new BulkCommitBatch(this.firestore);
-
-    // Send the batch if it is under the rate limit, or schedule another
-    // attempt after the appropriate timeout.
-    const underRateLimit = this._rateLimiter.tryMakeRequest(
+    // Use the write with the highest failure count when determining backoff.
+    const highestBackoffDuration = pendingBatch.pendingOps.reduce((prev, cur) =>
```

Comment:
I think this doesn't quite work. Let's say we are supposed to back off for 10 seconds, but the rate limiter tells us that we need to wait 20 seconds. Once 20 seconds passed, we should be able to send the request right away. If I read this correctly, we still apply the 10 second backoff here.

I think you need to use a "target time" rather than a duration.

Reply:
I think the backoffs are applied together. We take apply the `max()` of the backoff and ratelimiter delays, and call `bulkCommit()` on the batch immediately after. Where do we apply the 2nd backoff?

Reply:
If I read this correctly, if a batch hits a rate limit, the function call to compute the next backoff is executed again. This once again looks at the rate limit backoff (which would be 0) and re-computes the retry backoff. It looks like this last step starts from scratch again - instead, it should deduct the backoff that has already passed. Let me know if I am confused and this works differently.

Reply:
Discussed it on gvc. The current implementation doesn't apply double backoff, and can be merged. However, I still need to add rate limiter checking before calling `bulkCommit()`, which I will do in a separate PR on top of this one.

Reply:
After discussing offline, it looks like my understanding of how this worked was slightly off. Can stay as is.